src/cache/pathValueMerge.js

```diff
@@ -44,70 +44,49 @@ module.exports = function pathValueMerge(cache, pathValue) {
     };
 };
 
-function innerPathValueMerge(cache, pathValue) {
-    var path = pathValue.path;
-    var curr = cache;
-    var next, key, cloned, outerKey, iteratorNote;
-    var i, len;
-
-    for (i = 0, len = path.length - 1; i < len; ++i) {
-        outerKey = path[i];
-
-        // Setup the memo and the key.
-        if (outerKey && typeof outerKey === 'object') {
-            iteratorNote = {};
-            key = iterateKeySet(outerKey, iteratorNote);
-        } else {
-            key = outerKey;
-            iteratorNote = false;
-        }
-
-        do {
-            next = curr[key];
+/* eslint-disable no-constant-condition */
+function innerPathValueMerge(root, node, value, path, depth, reference) {
 
-            if (!next) {
-                next = curr[key] = {};
-            }
+    var note = {};
+    var branch = depth < path.length - 1;
+    var keySet = path[depth];
+    var key = iterateKeySet(keySet, note);
 
-            if (iteratorNote) {
-                innerPathValueMerge(
-                    next, {
-                        path: path.slice(i + 1),
-                        value: pathValue.value
-                    });
-
-                if (!iteratorNote.done) {
-                    key = iterateKeySet(outerKey, iteratorNote);
-                }
-            }
-
-            else {
-                curr = next;
-            }
-        } while (iteratorNote && !iteratorNote.done);
+    do {
+        var curr = node;
+        var type = curr.$type;
 
-        // All memoized paths need to be stopped to avoid
-        // extra key insertions.
-        if (iteratorNote) {
-            return;
+        while (type === $ref) {
+            curr = innerPathValueMerge(root, root, value, curr.value, 0, true);
+            type = curr.$type;
         }
-    }
-
 
-    // TODO: This clearly needs a re-write.  I am just unsure of how i want
-    // this to look.  Plus i want to measure performance.
-    outerKey = path[i];
+        if (type === void 0) {
```

Comment:
no `void 0` do `undefined` for explicitness sake considering void actually evaluates the `0` as an expression, executes it, and ignores the output, but instead returns undefined.